File: matlab2/create_timevary_stiff.py
import numpy as np
    
def create_timevary_stiff(Ugt = None,transmat = None,nodes = None,pars = None): 
    #计算时变刚度矩阵
#形函数:单元坐标系下节点自由度顺序为：z,x,ry,y,rx
    phix = lambda z = None,le = None: np.array([[0],[1 - 3 * z ** 2 / le ** 2 + 2 * z ** 3 / le ** 3],[z - 2 * z ** 2 / le + z ** 3 / le ** 2],[0],[0],[0],[3 * z ** 2 / le ** 2 - 2 * z ** 3 / le ** 3],[- z ** 2 / le + z ** 3 / le ** 2],[0],[0]])
    phiy = lambda z = None,le = None: np.array([[0],[0],[0],[1 - 3 * z ** 2 / le ** 2 + 2 * z ** 3 / le ** 3],[- (z - 2 * z ** 2 / le + z ** 3 / le ** 2)],[0],[0],[0],[3 * z ** 2 / le ** 2 - 2 * z ** 3 / le ** 3],[- (- z ** 2 / le + z ** 3 / le ** 2)]])
    phiz = lambda z = None,le = None: np.array([[1 - z / le],[0],[0],[0],[0],[z / le],[0],[0],[0],[0]])
    #采用差分方式定义一阶微分,二阶微分
    scale = 0.01
    
    dphix = lambda z = None,le = None: (phix(z + scale,le) - phix(z - scale,le)) / 2 / scale
    ddphix = lambda z = None,le = None: (dphix(z + scale,le) - dphix(z - scale,le)) / 2 / scale
    dphiy = lambda z = None,le = None: (phiy(z + scale,le) - phiy(z - scale,le)) / 2 / scale
    ddphiy = lambda z = None,le = None: (dphiy(z + scale,le) - dphiy(z - scale,le)) / 2 / scale
    dphiz = lambda z = None,le = None: (phiz(z + scale,le) - phiz(z - scale,le)) / 2 / scale
    ddphiz = lambda z = None,le = None: (dphiz(z + scale,le) - dphiz(z - scale,le)) / 2 / scale
    nelem = sum(pars.mesh)
    Kggadd = np.zeros(((nelem + 1) * 5,(nelem + 1) * 5))
    for i in np.arange(1,nelem+1).reshape(-1):
        Kadd = create_elem_timevary_stiff_in_local(nodes(i,:),nodes(i + 1,:),pars,phix,phiy,phiz,dphix,dphiy,dphiz,ddphix,ddphiy,ddphiz,Ugt(np.arange(i * 5 - 4,i * 5 + 5+1)),transmat[i])
        #组装附加刚度矩阵
        Kggadd[np.arange[i * 5 - 4,i * 5 + 5+1],np.arange[i * 5 - 4,i * 5 + 5+1]] = Kggadd(np.arange(i * 5 - 4,i * 5 + 5+1),np.arange(i * 5 - 4,i * 5 + 5+1)) + np.transpose(transmat[i]) * Kadd * transmat[i]
    
    return Kggadd
    
    
def create_elem_timevary_stiff_in_local(node1 = None,node2 = None,pars = None,phix = None,phiy = None,phiz = None,dphix = None,dphiy = None,dphiz = None,ddphix = None,ddphiy = None,ddphiz = None,Uet = None,elem_transmat = None): 
    
    #计算一些参数
    le = norm(node2(np.arange(1,3+1)) - node1(np.arange(1,3+1)),2)
    
    A1 = np.pi * node1(8) ** 2 - np.pi * node1(7) ** 2
    A2 = np.pi * node2(8) ** 2 - np.pi * node2(7) ** 2
    
    I1 = np.pi * node1(8) ** 4 / 4 - np.pi * node1(7) ** 4 / 4
    I2 = np.pi * node2(8) ** 4 / 4 - np.pi * node2(7) ** 4 / 4
    
    I = lambda z = None: interp1(np.array([0,le]),np.array([I1,I2]),z)
    
    A = lambda z = None: interp1(np.array([0,le]),np.array([A1,A2]),z)
    
    uet = elem_transmat * Uet
    
    K5 = quadv(lambda z = None: pars.Ev * A(z) * dphix(z,le) * np.transpose(uet) * (dphiz(z,le)) * np.transpose((phix(z,le))),0,le)
    
    K6 = quadv(lambda z = None: pars.Ev * A(z) * dphiy(z,le) * np.transpose(uet) * (dphiz(z,le)) * np.transpose((dphiy(z,le))),0,le)
    
    K7 = quadv(lambda z = None: 1 / 2 * pars.Ev * A(z) * dphix(z,le) * np.transpose((dphix(z,le))) * uet * np.transpose(uet) * phix(z,le) * np.transpose((phix(z,le))),0,le)
    
    K8 = quadv(lambda z = None: 1 / 2 * pars.Ev * A(z) * dphiy(z,le) * np.transpose((dphiy(z,le))) * uet * np.transpose(uet) * dphiy(z,le) * np.transpose((dphiy(z,le))),0,le)
    
    K9 = quadv(lambda z = None: 1 / 2 * pars.Ev * A(z) * dphix(z,le) * np.transpose((dphix(z,le))) * uet * np.transpose(uet) * dphiy(z,le) * np.transpose((dphiy(z,le))),0,le)
    
    K10 = quadv(lambda z = None: 1 / 2 * pars.Ev * A(z) * dphiy(z,le) * np.transpose((dphiy(z,le))) * uet * np.transpose(uet) * dphix(z,le) * np.transpose((dphix(z,le))),0,le)
    
    K12 = quadv(lambda z = None: 1 / 2 * pars.Ev * A(z) * dphiz(z,le) * np.transpose((dphix(z,le))) * uet * np.transpose((dphix(z,le))),0,le)
    
    K13 = quadv(lambda z = None: 1 / 2 * pars.Ev * A(z) * dphiz(z,le) * np.transpose((dphiy(z,le))) * uet * np.transpose((dphiy(z,le))),0,le)
    
    Kadd = K5 + K6 + K7 + K8 + K9 + K10 + K12 + K13
    return Kadd
    
    return Kggadd

File: matlab2/create_timevary_stiff.m
function Kggadd=create_timevary_stiff(Ugt,transmat,nodes,pars)
    %计算时变刚度矩阵
    %形函数:单元坐标系下节点自由度顺序为：z,x,ry,y,rx
    phix=@(z,le)[0;
               1-3*z^2/le^2+2*z^3/le^3;
               z-2*z^2/le+z^3/le^2;
               0;
               0;
               0;
               3*z^2/le^2-2*z^3/le^3;
               -z^2/le+z^3/le^2;
               0;...
               0];
   phiy=@(z,le)[0;
             0;
             0;
             1-3*z^2/le^2+2*z^3/le^3;
             -(z-2*z^2/le+z^3/le^2);
             0;
             0;
             0;
             3*z^2/le^2-2*z^3/le^3;
             -(-z^2/le+z^3/le^2)];
    phiz=@(z,le)[1-z/le;0;0;0;0;z/le;0;0;0;0];
    %采用差分方式定义一阶微分,二阶微分
    scale=0.01;         %差分尺度
    dphix=@(z,le)(phix(z+scale,le)-phix(z-scale,le))/2/scale;
    ddphix=@(z,le)(dphix(z+scale,le)-dphix(z-scale,le))/2/scale;
    dphiy=@(z,le)(phiy(z+scale,le)-phiy(z-scale,le))/2/scale;
    ddphiy=@(z,le)(dphiy(z+scale,le)-dphiy(z-scale,le))/2/scale;
    dphiz=@(z,le)(phiz(z+scale,le)-phiz(z-scale,le))/2/scale;
    ddphiz=@(z,le)(dphiz(z+scale,le)-dphiz(z-scale,le))/2/scale;
    nelem=sum(pars.mesh);
    Kggadd=zeros((nelem+1)*5,(nelem+1)*5);
    for i=1:nelem
        Kadd=create_elem_timevary_stiff_in_local(nodes(i,:),nodes(i+1,:),pars,...
            phix,phiy,phiz,dphix,dphiy,dphiz,ddphix,ddphiy,ddphiz,Ugt(i*5-4:i*5+5),transmat{i});
        %组装附加刚度矩阵
        Kggadd(i*5-4:i*5+5,i*5-4:i*5+5)=Kggadd(i*5-4:i*5+5,i*5-4:i*5+5)+transmat{i}'*Kadd*transmat{i};
    end
end

function Kadd=create_elem_timevary_stiff_in_local(node1,node2,pars,...
            phix,phiy,phiz,dphix,dphiy,dphiz,ddphix,ddphiy,ddphiz,Uet,elem_transmat)
    %
    %计算一些参数
    le=norm(node2(1:3)-node1(1:3),2);           %单元长度
    A1=pi*node1(8)^2-pi*node1(7)^2;
    A2=pi*node2(8)^2-pi*node2(7)^2;         %单元两个节点对应截面面积
    I1=pi*node1(8)^4/4-pi*node1(7)^4/4;
    I2=pi*node2(8)^4/4-pi*node2(7)^4/4;     %两个截面的惯性矩
    I=@(z)interp1([0,le],[I1,I2],z);        %油管惯性矩定义为函数
    A=@(z)interp1([0,le],[A1,A2],z);        %油管截面积定义为函数
    uet=elem_transmat*Uet;                  %单元坐标系下单元位移向量
    K5=quadv(@(z)pars.Ev*A(z)*dphix(z,le)*uet'*(dphiz(z,le))*(phix(z,le))',0,le);       %单元刚度矩阵5
    K6=quadv(@(z)pars.Ev*A(z)*dphiy(z,le)*uet'*(dphiz(z,le))*(dphiy(z,le))',0,le);       %单元刚度矩阵6
    K7=quadv(@(z)1/2*pars.Ev*A(z)*dphix(z,le)*(dphix(z,le))'*uet*uet'*phix(z,le)*(phix(z,le))',0,le);   %单元刚度矩阵7
    K8=quadv(@(z)1/2*pars.Ev*A(z)*dphiy(z,le)*(dphiy(z,le))'*uet*uet'*dphiy(z,le)*(dphiy(z,le))',0,le); %单元刚度矩阵8
    K9=quadv(@(z)1/2*pars.Ev*A(z)*dphix(z,le)*(dphix(z,le))'*uet*uet'*dphiy(z,le)*(dphiy(z,le))',0,le); %单元刚度矩阵9
    K10=quadv(@(z)1/2*pars.Ev*A(z)*dphiy(z,le)*(dphiy(z,le))'*uet*uet'*dphix(z,le)*(dphix(z,le))',0,le);%单元刚度矩阵10
    K12=quadv(@(z)1/2*pars.Ev*A(z)*dphiz(z,le)*(dphix(z,le))'*uet*(dphix(z,le))',0,le);%单元刚度矩阵12
    K13=quadv(@(z)1/2*pars.Ev*A(z)*dphiz(z,le)*(dphiy(z,le))'*uet*(dphiy(z,le))',0,le);%单元刚度矩阵13
    Kadd=K5+K6+K7+K8+K9+K10+K12+K13;
    
end
